Models docstrings

A-Jacobson opened the pull request:
#401 
- not sure how far i should go with the `loss.py`, `initializers.py` and model architecture files. The first two probably need a good refactor, loss -> metrics and initializers -> (not sure, maybe deletion). The model architectures are copied from all over the place. Am adding docstrings but am reluctant to refactor. Our `ComposerModel` versions should be the public facing interface for these. 

- unsure about module level docstring, should i do something like this? https://docs.mosaicml.com/en/v0.3.1/models.html. or link to the incoming model cards that @ajaysaini725  is writing to cover the model descriptions.

ravi-mosaicml commented:
> * not sure how far i should go with the `loss.py`, `initializers.py` and model architecture files. The first two probably need a good refactor, loss -> metrics and initializers -> (not sure, maybe deletion). The model architectures are copied from all over the place. Am adding docstrings but am reluctant to refactor. Our `ComposerModel` versions should be the public facing interface for these.

If a refactor is desperately needed, then let's do it. For PR review, though, could this refactor be done separately?


> * unsure about module level docstring, should i do something like this? https://docs.mosaicml.com/en/v0.3.1/models.html. or link to the incoming model cards that @ajaysaini725  is writing to cover the model descriptions.

A stub that would link to the model cards works.

hanlint commented:
Thanks for putting this together! A few comments:

`ComposerModel`:
- [ ] Can you add more information to the `ComposerModel` base class docstrings? e.g. especially the coupling of forward/loss and validate/metrics.
- [ ] `forward`: its worth mentiong that the input is from the dataloader. we want to tell folks as much about our trainer loop logic as possible. same thing for loss -- say that outputs come from forward, and batch from the dataloader
- [ ] `metrics`: similarly explain that the metrics shuld be torchmetrics. The "warning" should hide the torchmetrics link (e.g. See torchmetrics[link] for more details). Mention that one can return multiple metrics that would all be evaluated on the eval_dataloader.

- [] `ComposerTransformer` docstring should also be expanded with more information about how it works and example usages
- [] same for TIMM, BERT, ResNet, GPTmodel, etc, we need some code blocks that show its usage, and more detail on how it works.
- [] TIMM has some formatting issues with `testcode` and also the `model_name` and links in the parameters. See: https://mosaicml-composer--469.com.readthedocs.build/en/469/api_reference/composer.models.timm.model.html#composer.models.timm.model.Timm

hanlint commented:
This is ready for a re-review: cc: @hanlint @ravi-mosaicml

ravi-mosaicml commented:
I took a quick look at https://mosaicml-composer--469.com.readthedocs.build/en/469/api_reference/composer.models.html and noticed that many model-level docstrings and `__all__` are not defined. In addition, many references are not auto-linking (set `nitpicky = True` in `conf.py` to get the full list.

Let me know when I should take another look.

ravi-mosaicml reviewed (changes requested):
Requesting changes for module-level docstrings, `__all__`, and another pass on fixing broken references.

A-Jacobson commented:
> I took a quick look at https://mosaicml-composer--469.com.readthedocs.build/en/469/api_reference/composer.models.html and noticed that many model-level docstrings and `__all__` are not defined. In addition, many references are not auto-linking (set `nitpicky = True` in `conf.py` to get the full list.
> 
> Let me know when I should take another look.

There are many objects in here I don't want to expose to a user such as Hparams and model objects copied from other repos that our `composermodels` wrap. I didn't write docstrings for these. I don't think I can exclude them with `__all__` either because we export them for our model registery in `trainer_hparams`. Do you have a suggestion for what to do about these?

ravi-mosaicml commented:
> There are many objects in here I don't want to expose to a user such as Hparams and model objects copied from other repos that our `composermodels` wrap. I didn't write docstrings for these. I don't think I can exclude them with `__all__` either because we export them for our model registery in `trainer_hparams`. Do you have a suggestion for what to do about these?

Hparams should be exposed; #558 helped format how these are documented. Every field in an hparam class should be listed under the Params section of the docstring, but for each field, it is fine to have a one-liner that points to the underlying class. See https://docs.mosaicml.com/en/latest/api_reference/composer.callbacks.callback_hparams.html#module-composer.callbacks.callback_hparams as an example.

For model objects from other repos, are they imports, or did we copy and paste the code? Imports need not be listed in `__all__`, since one should import from the source package (not composer). If we copied and pasted the code, then we are responsible for the code now, and as such, it should have docstrings and be brought up to our standards.

A-Jacobson commented:
> Hparams should be exposed; #558 helped format how these are documented. Every field in an hparam class should be listed under the Params section of the docstring, but for each field, it is fine to have a one-liner that points to the underlying class. See https://docs.mosaicml.com/en/latest/api_reference/composer.callbacks.callback_hparams.html#module-composer.callbacks.callback_hparams as an example.
> 
- Exposed and documented Hparams.

> For model objects from other repos, are they imports, or did we copy and paste the code? Imports need not be listed in `__all__`, since one should import from the source package (not composer). If we copied and pasted the code, then we are responsible for the code now, and as such, it should have docstrings and be brought up to our standards.

- Much of the code is copy and pasted and I suspect a some doesn't need to be here. I gave the main models docstrings and set some helper files to private. I'm going to have to do a larger refactor PR to bring all the copied code up to our standards.. which I do have planned. Hopefully this will do for a start.

- defined __all__ in every module + public file.

- fixed many broken links, there are still some that are broken when i run `nitpicky = True` locally. The broken ones are all external modules such as `torchmetrics.collections.MetricCollection`, builtin typing objections, torch, torchvision, and transformers objects. Are these supposed to work? I see these linked in our sphinx config. @ravi-mosaicml

ravi-mosaicml commented:
> * fixed many broken links, there are still some that are broken when i run `nitpicky = True` locally. The broken ones are all external modules such as `torchmetrics.collections.MetricCollection`, builtin typing objections, torch, torchvision, and transformers objects. Are these supposed to work? I see these linked in our sphinx config. @ravi-mosaicml

Can you look into the internsphinx inventory and see what the proper path for these elements are? See https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#showing-all-links-of-an-intersphinx-mapping-file. OK to defer this one to later -- if so please create a seperate issue for it.

> * The `forward` method of `ComposerModel` isn't showing in the docs and I'm not sure why. Perhaps it's because we subclass `torch.nn.Module`?

Forward is excluded in the sphinx config. Would adding a `:meta public:` force it to be documented?

hanlint commented on composer/models/__init__.py:
```diff
@@ -1,5 +1,13 @@
+:class:`~composer.models.base.ComposerModel` s that wrap existing pytorch models with standard forward passes
+and logging to enable quick interaction with the :class:`~composer.trainer.trainer.Trainer`.
+
+See the ComposerModel docs for more details.
```
Should like to the docs with `:doc:` here.

hanlint commented on composer/models/base.py:
```diff
@@ -17,38 +18,125 @@
 
 class ComposerModel(torch.nn.Module, abc.ABC):
-    """The minimal interface needed to use a model with :class:`composer.trainer.Trainer`."""
+    """The interface needed to use a pytorch model with :class:`~composer.trainer.trainer.Trainer`.
```
```suggestion
    """The interface needed to use a pytorch model with :class:`.Trainer`.
```

hanlint commented on composer/models/base.py:
```diff
@@ -17,38 +18,125 @@
-    """The minimal interface needed to use a model with :class:`composer.trainer.Trainer`."""
+    """The interface needed to use a pytorch model with :class:`~composer.trainer.trainer.Trainer`.
+
+    To create a :class:`~composer.trainer.trainer.Trainer` compatible model, subclass :class:`~composer.models.base.ComposerModel` and
```
```suggestion
    To create a :class:`.Trainer` compatible model, subclass :class:`~composer.models.base.ComposerModel` and
```

hanlint commented on composer/models/base.py:
```diff
@@ -17,38 +18,125 @@
+            def loss(self, outputs, batch):
+                # pass batches and `forward` outputs to the loss
+                _, targets = batch
+                return F.cross_entropy(outputs, targets)
```
Should additionally add the same as above -> For more detail, see the composer model docs (with a :doc: link)

hanlint commented on composer/models/base.py:
```diff
@@ -17,38 +18,125 @@
+        """Compute model output given a batch from the dataloader.
+
+        .. warning:: this is method is different from vanilla pytorch `model.forward(x)` or `model(x)` as it takes a
+                     batch of data that may have to be unpacked.
```
```suggestion
                     batch of data that has to be unpacked.
```

hanlint commented on composer/models/base.py:
```diff
@@ -17,38 +18,125 @@
 
     def metrics(self, train: bool = False) -> Metrics:
-        """Get metrics for evaluating the model.
+        """Get metrics for evaluating the model. Metrics should be torchmetrics compatible for accurate distributed
```
```suggestion
        """Get metrics for evaluating the model. Metrics should be :class:`torchmetrics.Metrics` compatible for accurate distributed
```

hanlint commented on composer/models/base.py:
```diff
@@ -58,39 +146,72 @@ def metrics(self, train: bool = False) -> Metrics:
+            def validate(self, batch): # batch is the output of the dataloader
+                inputs, targets = batch
+                outputs = self.model(inputs)
+                return outputs, targets # return a tuple of (``outputs``, ``targets``)
```
```suggestion
                return outputs, targets # return a tuple of (outputs, targets)
```

code block, not RST

hanlint commented on composer/models/base.py:
```diff
@@ -58,39 +146,72 @@ def metrics(self, train: bool = False) -> Metrics:
+                return outputs, targets # return a tuple of (``outputs``, ``targets``)
+
+
+        How validate outputs are passed to :meth:`metrics` (pseudocode):
```
```suggestion
        This pseudocode illustrates how validate outputs are passed to :meth:`metrics`:
```

hanlint commented on composer/models/base.py:
```diff
@@ -58,39 +146,72 @@ def metrics(self, train: bool = False) -> Metrics:
 
 class ComposerClassifier(ComposerModel):
-    """Implements the base logic that all classifiers can build on top of.
+    """A convenience class that creates a :class:`~composer.models.base.ComposerModel` for classification tasks from a
```
here and throughout should use the global scope `.ComposerModel` and `.ComposerClassifier` to make the docstrings more less verbose.

hanlint commented on composer/models/bert/model.py:
```diff
@@ -96,13 +108,12 @@ def validate(self, batch: BatchDict) -> Tuple[Tensors, Tensors]:
         Args:
             batch (BatchDict): a dictionary of Dict[str, Tensor] of inputs
-                that the model expects, as found in ComposerTransformer.get_model_inputs().
+                that the model expects, as found in :meth:`composer.models.transformer_shared.ComposerTransformer.get_model_inputs`.
```
```suggestion
                that the model expects, as found in :meth:`.ComposerTransformer.get_model_inputs`.
```

hanlint commented on composer/models/classify_mnist/model.py:
```diff
@@ -47,16 +49,28 @@ def forward(self, x):
 class MNIST_Classifier(ComposerClassifier):
     """A simple convolutional neural network extending :class:`ComposerClassifier`.
```
```suggestion
    """A simple convolutional neural network extending :class:`.ComposerClassifier`.
```

hanlint commented on composer/models/classify_mnist/model.py:
```diff
@@ -47,16 +49,28 @@ def forward(self, x):
     """A simple convolutional neural network extending :class:`ComposerClassifier`.
 
-    :class:`composer.models.MNIST_Classifier` is a simple example
+    :class:`~composer.models.classify_mnist.model.MNIST_Classifier` is a simple example
```
```suggestion
    A simple example
```
(self-referencing)

hanlint commented on composer/models/deeplabv3/deeplabv3.py:
```diff
@@ -64,15 +67,24 @@ def deeplabv3_builder(num_classes: int,
 
 class ComposerDeepLabV3(ComposerModel):
-    """DeepLabV3 model extending the :class:`ComposerClassifier`.
+    """DeepLabV3 model extending :class:`~composer.models.base.ComposerClassifier`. Logs Mean Intersection over Union
```
```suggestion
    """DeepLabV3 model extending :class:`.ComposerClassifier`. Logs Mean Intersection over Union
```

hanlint commented on composer/models/deeplabv3/deeplabv3_hparams.py:
```diff
@@ -6,10 +6,13 @@
 @dataclass
 class DeepLabV3Hparams(ModelHparams):
-    """This class specifies arguments for a DeepLabV3 model and can instantiate a DeepLabV3 model.
+    """yahp interface for :class:`~composer.models.deeplabv3.deeplabv3.ComposerDeepLabV3`. This class specifies
```
```suggestion
    """yahp interface for :class:`~.ComposerDeepLabV3`. This class specifies
```

hanlint reviewed (commented):
This PR is a lot, thanks for doing this. Made several comments. Some major comments are:

* for brevity in the code, please use the global scope e.g. `.ComposerModel` instead of the full path `~composer....`. 
* defaults should be formatted as Default: True, not Default = True.

A-Jacobson commented:
> This PR is a lot, thanks for doing this. Made several comments. Some major comments are:
> 
> * for brevity in the code, please use the global scope e.g. `.ComposerModel` instead of the full path `~composer....`.
> * defaults should be formatted as Default: True, not Default = True.

Thanks for going through it. Comments should be addressed.

hanlint reviewed (approved):
Agreed that the `loss` and `initializers` will need a better organization moving forward (already captured in #536 ), but that should not be blocking for this PR.

However, one major request -- for all docstrings, arguments that have defaults should be listed as optional. eg. in `Timm`, should be:

```
pretrained (bool, optional): Imagenet pretrained. Default: False.
```
Can you do a pass through all the methods in this module and add optional wherever appropriate?

growlix commented on composer/models/unet/__init__.py:
```diff
@@ -1,8 +1,15 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+"""The Unet architecture used in image segmentation. The example we are using is for medical brain tumor data.
```
The 2nd sentence could be a bit more specific. The example dataset?

growlix commented on composer/models/unet/model.py:
```diff
@@ -3,10 +3,25 @@
+        n_class (int): number of output layers.
+        kernels (list): conv layer kernel sizes.
+        strides (list): conv layer strides.
+        normalization_layer (str): normalization layer type, one of ("batch", "instance").
```
```suggestion
        normalization_layer (str): normalization layer type, one of (``"batch"``, ``"instance"``).
```

growlix commented on composer/models/unet/model.py:
```diff
@@ -3,10 +3,25 @@
+    https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Segmentation/nnUNet/.
+
+    Args:
+        in_channels (int): number of input channels.
+        n_class (int): number of output layers.
+        kernels (list): conv layer kernel sizes.
+        strides (list): conv layer strides.
+        normalization_layer (str): normalization layer type, one of ("batch", "instance").
+        negative_slope (float): leaky relu negative slope.
+        residual (bool): use residual connections.
+        dimension (int): filter dimensions.
```
Capitalize first letter of parameter description

growlix commented on composer/models/unet/model.py:
```diff
@@ -3,10 +3,25 @@
 ## Code adapted from https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Segmentation/nnUNet/
```
File-level docstring

growlix commented on composer/models/unet/unet.py:
```diff
@@ -1,6 +1,5 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
```
File-level docstring

growlix commented on composer/models/unet/unet.py:
```diff
@@ -15,16 +14,20 @@
 
-    n_classes: Optional[int] = None
+    Args:
+        num_classes (int): The number of classes. Needed for classification tasks. Default: 3.
```
```suggestion
        num_classes (int): The number of classes. Needed for classification tasks. Default: ``3``.
```

growlix commented on composer/models/unet/unet.py:
```diff
@@ -15,16 +14,20 @@
 
-    See this `paper <https://arxiv.org/abs/1505.04597>`_ for details on the U-Net architecture.
-    """
+    See this `paper <https://arxiv.org/abs/1505.04597>`_ for details on the
```
Reference format

growlix commented on composer/models/unet/model.py:
```diff
@@ -3,10 +3,25 @@
 
 
 class UNet(nn.Module):
+    """Unet Architecture adapted from
+    https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Segmentation/nnUNet/.
```
embed link

growlix commented on composer/models/unet/unet_hparams.py:
```diff
@@ -1,13 +1,20 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
```
File-level docstring

growlix commented on composer/models/unet/unet_hparams.py:
```diff
@@ -1,13 +1,20 @@
+    """yahp Hparams interface for Unet segmentation models.
+
+    Args:
+        num_classes (int, optional): The number of classes. Needed for classification tasks. Default: 3.
```
```suggestion
        num_classes (int, optional): The number of classes. Needed for classification tasks. Default: ``3``.
```

growlix reviewed (commented):
99% of my comments are one of the following four comments:

1. Example and default arguments should always be formatted like ``"this"``.
2. Consistent reference formatting
3. No nude hyperlinks—embed hyperlinks in descriptive text
4. File-level docstrings

I'm sorry that we have to inflict this violence on each other. But impeccable documentation requires blood sacrifice.

A-Jacobson commented on composer/models/base.py:
```diff
@@ -1,54 +1,150 @@
+
+        Example:
+
+        .. code-block:: python
```
Going to punt the doctests in the base class to another pr as they require fixtures.

A-Jacobson commented on composer/models/base.py:
```diff
@@ -1,54 +1,150 @@
     def metrics(self, train: bool = False) -> Metrics:
-        """Get metrics for evaluating the model.
+        """Get metrics for evaluating the model. Metrics should be instances of :class:`torchmetrics.Metric` defined in
+        :meth:`__init__`. This format enables accurate distributed logging. Metrics consume the outputs of
```
I'm referring to the sub class init here. Not sure how to link to that as it doesn't exist yet :(. Base class doesn't have an init.

A-Jacobson commented on composer/models/loss.py:
```diff
@@ -153,7 +159,39 @@ def soft_cross_entropy(input: Tensor,
+            in case of 2D Loss, or :math:`(N, C, d_1, d_2, ..., d_K)` where :math:`K \geq 1`
+            in the case of K-dimensional loss. `input` is expected to contain unnormalized scores
+            (often referred to as logits).
+        target (torch.Tensor) : If containing class indices, shape :math:`(N)` where each value is
```
It should! This is the pytorch docstring with some minor modifications

A-Jacobson commented on composer/models/base.py:
```diff
@@ -1,54 +1,150 @@
+
+    Minimal Example:
+
+    .. code-block:: python
```
gonna punt this to another PR, ravi and i discussed these.

A-Jacobson commented on composer/models/base.py:
```diff
@@ -1,54 +1,150 @@
         Args:
-            batch (Batch): The input batch for the forward pass.
+            outputs (Any): The output of the forward pass.
+            batch (~composer.core.types.Batch): The output batch from dataloader.
```
have tried a few things to fix this. was unable to. Also these are going to be removed soon anyway so I didn't want to burn too much time on it.

A-Jacobson commented on composer/models/deeplabv3/deeplabv3.py:
```diff
@@ -22,10 +24,11 @@ def deeplabv3_builder(num_classes: int,
 
```
beyond    

 """Helper function to build a torchvision DeepLabV3 model with a 3x3 convolution layer and dropout removed."""
 
 ?

A-Jacobson commented on composer/models/deeplabv3/deeplabv3.py:
```diff
@@ -64,15 +67,25 @@ def deeplabv3_builder(num_classes: int,
 
 
 class ComposerDeepLabV3(ComposerModel):
```
only that I didn't want to add a refactor to this already huge PR. see https://github.com/mosaicml/composer/issues/536

A-Jacobson commented on composer/models/base.py:
```diff
@@ -58,39 +154,72 @@ def metrics(self, train: bool = False) -> Metrics:
-                `update()` methods of the metrics returned by :meth:`metrics`.
-                Most often, this will be a tuple of the form (predictions, targets).
+            Tuple[Any, Any]: A Tuple of (``outputs``, ``targets``) that is passed directly to the
+                :meth:`~torchmetrics.Metric.update` methods of the metrics returned by :meth:`metrics`.
```
yeah a lot of torchmetrics links are broken. The package authors don't have working links of it either. https://torchmetrics.readthedocs.io/en/latest/references/modules.html?highlight=Metrics . not sure what to do about this.

A-Jacobson commented on composer/models/bert/model.py:
```diff
@@ -19,16 +19,27 @@
-        config (transformers.BertConfig): The config for the model.
-        tokenizer (transformers.BertTokenizer): The tokenizer used for this model,
-            necessary to assert required model inputs.
+        module (transformers.BertModel): An instance of BertModel that
```
same as torchmetrics, I believe transformers doesn't have a compatible RTD table i'm not sure how to get them to work.

A-Jacobson commented on composer/models/nlp_metrics.py:
```diff
@@ -78,7 +81,7 @@ def update(self, output: Union[Mapping, Tensor], target: Tensor) -> None:
         Args:
             output (Mapping): The output from the model, which must contain
```
Have tried a few path variations with no luck. any ideas on this?

A-Jacobson commented on composer/models/nlp_metrics.py:
```diff
@@ -54,9 +57,9 @@ class CrossEntropyLoss(Metric):
-        ignore_index (int): The class index to ignore. Defaults to -100.
+        ignore_index (int, optional): The class index to ignore. Defaults to -100.
 
     State:
```
saw the slack discussion on this but it doesn't look like there's a great solution. either set as attributes or ignore it for now?

A-Jacobson commented:
Done:

- Added all file level docstrings.
- embedding links.
- added all the `` marks...everywhere.
- vit docstrings, sorta minimal.
- bare minimum ssd docstrings.
- refactored effecientnet to hide non user-facing blocks.

Not Done:
- some links are still broken: transformers, base types, some torchmetrics. These broken links don't have readthedocs paths available for us to you. I've linked to the relevant documentation or github pages for some. For other's i've left alone either because we're working to remove them (Batch types), or they're in an awkward place to add a normal link (Return or Args). Open to suggestions here but I do think it's ok to send this off without them.
- codeblocks -> doctests in abstract classes, punting to a later PR as they require fixtures + some are psuedocode.
- `State` formatting on metrics, adding them as attrs will cause more trouble than living with the less pretty formatting for now.
- SSD is a lot.. I know it's not perfect but I'd like to get this PR moving before even more models appear.

@growlix I choose you!

growlix commented:
Looking greaet, Austin. Hope your eyes/fingers aren't too bloody.

> embedded paper links, they are cited as [Name](Link) not [Name](link)(Author et al, Date). Should i do (Author et al, Date)? Is there a relatively painless way to get this information?

I'd go [Name](link)(Author et al, Date) if possible; [Author et al, Date](link) is also acceptable. I think author attribution is the most important part. Unfortunately I don't know if there's an easy way to do this other than clicking through to the actual page.

> some links are still broken: transformers, base types, some torchmetrics. These broken links don't have readthedocs paths available for us. I've linked to the relevant documentation or github pages for some. I've left others alone either because we're working to remove them (Batch types), or they're in an awkward place to add a normal link (Return or Args). Open to suggestions here but I do think it's ok to send this off without them.

Agree. I think you can leave as-is.

> codeblocks -> doctests in abstract classes, punting to a later PR as they require fixtures + some are psuedocode.

Sounds good, just remember to file an issue.

> SSD is a lot.. I know it's not perfect but I'd like to get this PR moving before even more models appear.

I'll double check for anything egregious, but otherwise sounds good, file an issue.

> State formatting on metrics, adding them as attrs will cause more trouble than living with the less pretty formatting for now.

As per [Bandish's suggestion](https://mosaicml.slack.com/archives/C02F05H39SL/p1646773237565229?thread_ts=1646768731.433729&cid=C02F05H39SL), I think you can just move the state info to a new paragraph in the summary and preface it with something like

```
Adds the following attributes to :class:`composer.core.state`:
```
(Don't quote my syntax on that)

Unfortunately I think the current rendering is a bit too sloppy.

> general refactoring (standardizing module naming and layout) is being punted to here, https://github.com/mosaicml/composer/issues/536. This PR is already too crazy, lets end the pain.

+ 1

A-Jacobson commented:
- author attribution in.
- state information moved above params on `nlp_metrics` not sure this is prettier but it's done.
- issue filed for codeblocks -> testcode https://github.com/mosaicml/composer/issues/703
- issue filed for ssd cleanup https://github.com/mosaicml/composer/issues/704

growlix reviewed (approved):
SSD files need `__all__`, but otherwise looks p good

A-Jacobson commented on composer/models/ssd/ssd.py:
```diff
@@ -1,3 +1,5 @@
+"""Single Shot Object detection Model with pretrained ResNet34 backbone extending :class:`.ComposerModel`."""
```
im cool with it.
